File: packages/core/src/identity.ts
/**
 * Project/developer identity, per design doc §8 — zero setup ceremony by
 * construction. Every developer's CLI/daemon computes these identically and
 * independently; there is no registration step and nothing to keep in sync.
 */

import { execFileSync } from "node:child_process";
import * as fs from "node:fs";
import * as path from "node:path";
import * as crypto from "node:crypto";
import * as os from "node:os";

function git(args: string[], cwd: string): string | null {
  try {
    // stdio: pipe stdout (we read it), discard stderr -- every call site here
    // treats a git failure (no such remote, not a git repo, no commits yet)
    // as an expected, silent `null`, not a real error; without this override
    // git's own "fatal: ..." lands on our stderr anyway even though we
    // already handle the failure, e.g. a fresh repo with no `origin` remote
    // printing "fatal: No such remote 'origin'" on every getOriginRemoteUrl
    // call despite that being a normal, handled case (§17 Phase 3's no-remote
    // fallback).
    return execFileSync("git", args, { cwd, encoding: "utf8", stdio: ["ignore", "pipe", "ignore"] }).trim();
  } catch {
    return null;
  }
}

function readOrCreatePersistedId(idPath: string): string {
  if (fs.existsSync(idPath)) {
    const existing = fs.readFileSync(idPath, "utf8").trim();
    if (existing) return existing;
  }
  const generated = crypto.randomUUID();
  try {
    fs.mkdirSync(path.dirname(idPath), { recursive: true });
    fs.writeFileSync(idPath, generated);
  } catch {
    // Best-effort persistence; an ephemeral id for this run is still
    // correct, just not durable across processes.
  }
  return generated;
}

/**
 * Canonicalizes a git remote URL so equivalent clone forms hash to the same
 * `projectId` (§8) -- discovered live, 2026-08-11: two clones of the same
 * repo using `git@host:org/repo.git` (SSH) vs `https://host/org/repo.git`
 * computed *different* projectIds and never saw each other's claims or
 * designs at all. Must stay byte-for-byte equivalent to the Go port in
 * `hook/identity.go`'s `canonicalizeRemoteURL`, or this breaks the same way
 * again for whichever side drifts.
 */
export function canonicalizeRemoteUrl(url: string): string {
  let s = url.trim();

  // SCP-like SSH syntax: git@host:org/repo(.git) -> host/org/repo
  const scpMatch = s.match(/^[^@/]+@([^:/]+):(.+)$/);
  if (scpMatch) {
    s = `${scpMatch[1]}/${scpMatch[2]}`;
  } else {
    // scheme://[user@]host/path -> host/path (https, http, ssh, git, ...)
    s = s.replace(/^[a-z][a-z0-9+.-]*:\/\/(?:[^@/]+@)?/i, "");
  }

  s = s.replace(/\.git$/i, "");
  s = s.replace(/\/+$/, "");

  // Hosts are case-insensitive, and GitHub-style paths route
  // case-insensitively too -- lowercase the whole thing rather than risk a
  // false negative (missed cross-session match) from an incidental case
  // difference between two clones.
  return s.toLowerCase();
}

/** `projectId = sha256(canonicalized git remote get-url origin)` (§8). */
export function computeProjectId(repoRoot: string): string {
  const remoteUrl = git(["remote", "get-url", "origin"], repoRoot);
  if (remoteUrl) {
    return crypto.createHash("sha256").update(canonicalizeRemoteUrl(remoteUrl)).digest("hex");
  }
  // Edge case (§8): no remote means no way to clone, so no cross-developer
  // coordination need by construction — a random id gitignored per-repo.
  return readOrCreatePersistedId(path.join(repoRoot, ".git", "twing-project-id"));
}

/** Suggests a `developerId` from `git config user.email` -- used only as
 * the default label offered at `twing keygen`/`admin bootstrap` time, and
 * by `align.ts`'s no-server git-diff fallback path (which never talks to a
 * server to verify anything against). Once registered, `developerId` is
 * server-issued and resolved from the authenticated PAT on every request
 * (§17.10 hardening) -- this local computation is no longer trusted as-is
 * for anything the server persists or acts on. */
export function computeDeveloperId(repoRoot: string): string {
  const email = git(["config", "user.email"], repoRoot);
  if (email) return email;
  return readOrCreatePersistedId(path.join(os.homedir(), ".twing", "id"));
}

export function computeBranch(repoRoot: string): string {
  return git(["rev-parse", "--abbrev-ref", "HEAD"], repoRoot) ?? "unknown";
}

/** §17 Phase 3: the raw `origin` remote URL, exposed separately from
 * `computeProjectId` (which only ever needs the hashed/canonicalized form)
 * because `init.ts` needs the canonicalized-but-unhashed string to feed
 * `parseGithubOwnerRepo` below. `null` when there's no `origin` remote at
 * all -- same case `computeProjectId` falls back to a persisted random id
 * for. */
export function getOriginRemoteUrl(repoRoot: string): string | null {
  return git(["remote", "get-url", "origin"], repoRoot);
}

/**
 * §17 Phase 3: extracts `{owner, repo}` from a *canonicalized* remote URL
 * (i.e. already run through `canonicalizeRemoteUrl` above) when it's
 * GitHub-hosted, `undefined` otherwise -- a non-GitHub host, or the
 * no-remote random-id fallback case (never call this with anything but a
 * real canonicalized URL). Deliberately strict: exactly two path segments
 * after the host, or this returns `undefined` rather than guessing --
 * `canonicalizeRemoteUrl` already stripped `.git`/trailing slashes/case, so
 * a real GitHub repo URL always canonicalizes to exactly
 * `github.com/owner/repo`.
 */
export function parseGithubOwnerRepo(canonicalRemoteUrl: string): { owner: string; repo: string } | undefined {
  const match = canonicalRemoteUrl.match(/^github\.com\/([^/]+)\/([^/]+)$/);
  if (!match) return undefined;
  return { owner: match[1], repo: match[2] };
}

/** Combines `getOriginRemoteUrl` + `canonicalizeRemoteUrl` + `parseGithubOwnerRepo`
 * -- the one call every GitHub-verified-onboarding call site needs (`init.ts`'s
 * founding-time seed, and, as of the `init`-wired default join/found path,
 * `join.ts` and `init.ts`'s auth resolution too). `undefined` for a repo with
 * no `origin` remote or a non-GitHub one -- callers use that to skip the
 * GitHub path entirely (e.g. `init.ts` avoids ever popping the device-flow
 * browser prompt for a repo it can't possibly work on). */
export function githubBinding(repoRoot: string): { owner: string; repo: string } | undefined {
  const remoteUrl = getOriginRemoteUrl(repoRoot);
  return remoteUrl ? parseGithubOwnerRepo(canonicalizeRemoteUrl(remoteUrl)) : undefined;
}

File: hook/identity.go
package main

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"
)

// Go port of packages/core/src/identity.ts's computeProjectId/
// computeDeveloperId, for the §17 design-gate path only -- this hook binary
// deliberately has no shared code with the TS side (design doc §4), so the
// logic is duplicated here rather than imported.

func gitOutput(cwd string, args ...string) (string, bool) {
	cmd := exec.Command("git", args...)
	cmd.Dir = cwd
	out, err := cmd.Output()
	if err != nil {
		return "", false
	}
	return strings.TrimSpace(string(out)), true
}

func readOrCreatePersistedID(idPath string) string {
	if data, err := os.ReadFile(idPath); err == nil {
		if trimmed := strings.TrimSpace(string(data)); trimmed != "" {
			return trimmed
		}
	}
	buf := make([]byte, 16)
	_, _ = rand.Read(buf)
	generated := hex.EncodeToString(buf)
	// Write into the directory, never create it -- mirrors the same guard in
	// identity.ts's readOrCreatePersistedId. MkdirAll here conjured a `.git`
	// where there was none, leaving a directory that is not a repository but
	// looks exactly like one to every existsSync(".git") check, including
	// findRepoRoot's, which then stopped its walk there forever after. Found
	// on the TS side 2026-09-16 and fixed there; this port kept the bug until
	// 2026-09-18, when a stray $HOME/.git holding nothing but twing-project-id
	// was still being attributed to as a phantom project.
	if info, err := os.Stat(filepath.Dir(idPath)); err != nil || !info.IsDir() {
		// No durable home for it: an ephemeral id for this run is still
		// correct, just not stable across processes.
		return generated
	}
	_ = os.WriteFile(idPath, []byte(generated), 0o644)
	return generated
}

// generateGroupID mints a fresh, unpersisted random id (§17 design linking,
// 2026-08) -- same crypto/rand + hex.EncodeToString mechanic as
// readOrCreatePersistedID above, deliberately without the disk-persistence
// half: a plan's groupId only needs to stay stable for the duration of one
// ExitPlanMode invocation (reused across every matching candidate's
// registration call within that single pass), never across separate
// invocations -- see design-store.ts's reregisterFromPlan doc comment for
// why a *retried* plan safely discards a freshly-minted groupId here rather
// than needing one persisted.
func generateGroupID() string {
	buf := make([]byte, 16)
	_, _ = rand.Read(buf)
	return hex.EncodeToString(buf)
}

var scpLikeRemoteRe = regexp.MustCompile(`^[^@/]+@([^:/]+):(.+)$`)
var schemeRemoteRe = regexp.MustCompile(`(?i)^[a-z][a-z0-9+.-]*://(?:[^@/]+@)?`)

// canonicalizeRemoteURL mirrors core/identity.ts's canonicalizeRemoteUrl --
// must stay byte-for-byte equivalent, or cross-session detection breaks the
// same way it did in production, 2026-08-11 (SSH vs HTTPS clones of the
// same repo hashing to different projectIds).
func canonicalizeRemoteURL(raw string) string {
	s := strings.TrimSpace(raw)

	if m := scpLikeRemoteRe.FindStringSubmatch(s); m != nil {
		s = m[1] + "/" + m[2]
	} else {
		s = schemeRemoteRe.ReplaceAllString(s, "")
	}

	if strings.HasSuffix(strings.ToLower(s), ".git") {
		s = s[:len(s)-len(".git")]
	}
	s = strings.TrimRight(s, "/")

	return strings.ToLower(s)
}

// computeProjectID mirrors identity.ts's computeProjectId: sha256(canonicalized
// git remote get-url origin), falling back to a gitignored random id per repo
// (§8). `cwd` need not be the repo root -- git resolves the enclosing repo
// from any subdirectory.
//
// The fallback resolves that root explicitly rather than joining `cwd`, which
// is the second half of the phantom-.git fix above: a repo with no origin,
// entered from a subdirectory, used to persist its id at
// `<subdir>/.git/twing-project-id` -- a phantom nested *inside* a real repo,
// which findRepoRoot then stops at instead of the true root, so the same
// checkout reports two different projects depending on where the hook fired.
//
// Returning "" for a non-repository is this side's answer to the error
// identity.ts raises there. It cannot throw: a hook that fails a tool call is
// worse than one that declines to name a project (§4, and this module's own
// "always exits 0" contract). Every current caller resolves a coordinator
// first, which already requires a repo root, so this is a guard against a
// future caller rather than a live path -- but the id it would otherwise mint
// is a syntactically perfect name for a project that has never existed, and
// that is exactly how the TS side spent a month querying a phantom.
func computeProjectID(cwd string) string {
	if remoteURL, ok := gitOutput(cwd, "remote", "get-url", "origin"); ok && remoteURL != "" {
		sum := sha256.Sum256([]byte(canonicalizeRemoteURL(remoteURL)))
		return hex.EncodeToString(sum[:])
	}
	repoRoot, ok := gitOutput(cwd, "rev-parse", "--show-toplevel")
	if !ok || repoRoot == "" {
		return ""
	}
	return readOrCreatePersistedID(filepath.Join(repoRoot, ".git", "twing-project-id"))
}

// developerId is no longer computed on this side at all for the full-auth
// path (§17.10 hardening): the server resolves it from the authenticated
// bearer PAT, never from anything the hook could compute locally. The
// original `computeDeveloperID` was removed along with its one call site
// in design_gate.go for exactly that reason.
//
// computeDeveloperID below is its reintroduction, scoped to §17 Phase 4
// only: a --no-auth coordinator has no bearer token to resolve an identity
// from at all, so *something* self-declared has to travel as
// X-Twing-Developer-Id -- mirrors identity.ts's computeDeveloperId
// (git-email-derived, falling back to a persisted random id) exactly.
// Every call site must gate this behind config.NoAuth; it must never be
// called on the full-auth path.
func computeDeveloperID(cwd string) string {
	if email, ok := gitOutput(cwd, "config", "user.email"); ok && email != "" {
		return email
	}
	home, err := os.UserHomeDir()
	if err != nil {
		home = "."
	}
	return readOrCreatePersistedID(filepath.Join(home, ".twing", "id"))
}
